diff -Naur a/configure.ac b/configure.ac
--- a/configure.ac	2015-08-30 14:14:38.338017626 -0500
+++ b/configure.ac	2015-08-30 14:14:44.661042131 -0500
@@ -8,7 +8,6 @@
 
 AC_PATH_PROG([GUILE], [guile])
 AC_CONFIG_FILES([Makefile doc/Makefile examples/Makefile data/Makefile])
-AC_CONFIG_FILES([sly/config.scm])
 AC_CONFIG_FILES([sandbox], [chmod +x sandbox])
 AC_CONFIG_FILES([pre-inst-env], [chmod +x pre-inst-env])
 
diff -Naur a/Makefile.am b/Makefile.am
--- a/Makefile.am	2015-08-30 14:14:38.337017623 -0500
+++ b/Makefile.am	2015-08-30 14:14:44.661042131 -0500
@@ -20,6 +20,9 @@
 moddir=$(prefix)/share/guile/site/2.0
 godir=$(libdir)/guile/2.0/ccache
 
+sly/config.scm: sly/config.scm.in
+	sed 's|@pkgdatadir[@]|$(pkgdatadir)|g' sly/config.scm.in > sly/config.scm
+
 SOURCES = 					\
   sly/utils.scm					\
   sly/agenda.scm 				\
